### ISSUE TITLE
HDR output on Pi4

### ISSUE THREAD

**Issue by veldenb** [moonlight-stream/moonlight-qt#710]
HDR on the Pi4 is currently not enabled but seems to be available.

Playing HDR video's using libreELEC 10.0.1 on a Pi4 is working great for me. Because Moonlight is also using ffmpeg I wonder, could Moonlight also enable HDR output on the Pi4?

Forcing HDR using videocfg=3 in moonlight-config seems to enable HDR in the stream.
I can Toggle HDR in games and when HDR is switched off the image looks normal.
When I turn HDR on in a game the image looks washed out, the TV is not switching to an HDR signal.
So it looks like the stream is switching to HDR but the Pi4 and the TV are not switching to HDR mode.

Moonlight log contains the following line that (I think) indicates ffmpeg tries to process a 10 bit stream:
"00:01:21 - FFmpeg: [hevc @ 0x1c35440] Main 10 profile bitstream"

Maybe HDR can be enabled on the Pi4 by using the LibreELEC ffmpeg or recent [rpi ffmpeg patches](https://github.com/jc-kynesim/rpi-ffmpeg)?

I can debug/test if necessary.
 
**LibreELEC related stuff** 
[Release notes LibreELEC 10.0.1](https://libreelec.tv/2021/11/03/libreelec-matrix-10-0-1/) for the Pi4 state : 
- NEW: HDR output (HDR10 and HLG)

These are the changes between LibreELEC 10.0.0 and 10.0.1 on Github: [here](https://github.com/LibreELEC/LibreELEC.tv/compare/10.0.0...10.0.1)

**Client**
Hardware: Pi4
OS: LibreELEC 10.0.1 using plugin (https://github.com/veldenb/plugin.program.moonlight-qt)
Display: Philips OLED HDR capable TV

**Server**
Hardware: x86_64
OS: Windows 10 (HDR enabled)
Display: Dell HDR capable monitor
GeForce Experience: v3.24.0.126

**Moonlight log**
Full Moonlight log trying HDR on the Pi4: [moonlight.log](https://github.com/moonlight-stream/moonlight-qt/files/7857911/moonlight.log)

**Comment by cgutman** [comment]
It is quite likely possible via EGL and/or DRM renderers, but it won't work today due to the lack of HDR metadata propagation.

I plan to look at HDR stuff again after the Moonlight Qt v3.2.0 release is finished.

**Comment by veldenb** [comment]
That would be great :)

**Comment by cgutman** [comment]
This should be working in the latest code now. You'll need to be using the Full KMS driver and running from the console directly.

Give it a try and see if it works for you.

**Comment by veldenb** [comment]
Thanks, going to try it this weekend

**Comment by veldenb** [comment]
Are there extra steps involved for compiling on the rpi4? I compiled the master branch on Debian buster but no luck.
Looks like the hardware decoder is not recognised:

```
00:00:00 - Qt Info: Unable to detect Wayland or X11, so EGLFS will be used by default. Set QT_QPA_PLATFORM to override this.
00:00:00 - Qt Info: Setting display mode by default. Set QT_QPA_EGLFS_ALWAYS_SET_MODE=0 to override this.
00:00:00 - SDL Info (0): Compiled with SDL 2.0.9
00:00:00 - SDL Info (0): Running with SDL 2.0.9
00:00:00 - Qt Info: No translation available for  "C"
00:00:00 - Qt Warning: qrc:/gui/main.qml:12:1: QML ApplicationWindow: ToolTip must be attached to an Item
00:00:00 - Qt Debug: "No carrier"
00:00:10 - Qt Debug: "No carrier"
00:00:20 - Qt Debug: "No carrier"
00:00:30 - Qt Debug: "No carrier"
00:00:40 - Qt Debug: "No carrier"
00:00:50 - Qt Debug: "No carrier"
00:00:51 - Qt Info: Found "gamecontrollerdb.txt" at "/storage/.kodi/userdata/addon_data/plugin.program.moonlight-qt/moonlight-home/.cache/Moonlight Game Streaming Project/Moonlight/gamecontrollerdb.txt"
00:00:51 - SDL Info (0): Loaded 244 new gamepad mappings
00:00:51 - SDL Error (0): SDL_InitSubSystem(SDL_INIT_VIDEO) failed:
00:00:51 - Qt Fatal: ASSERT: "maximumStreamingFrameRate >= 60" in file backend/systemproperties.cpp, line 73
Aborted
```

**Comment by veldenb** [comment]
Tried a fresh native raspbian install (previous compile was Docker).
Starts but no HDR yet, logs:
[moonlight-hdr.log](https://github.com/moonlight-stream/moonlight-qt/files/7966733/moonlight-hdr.log)
Picture is also not fullscreen

**Comment by veldenb** [comment]
I deleted my previous 2 post since they didn't mention anything useful to this thread.
I managed to build the master branch using the rpi build-script from https://github.com/cgutman/moonlight-packaging.

I configured the hvec experimental HDR mode en started a HDR game but it looks like SDL isn't able to parse the HDR meta data from my tv.

Full moonlight-qt log: [moonlight-qt-hevs-hdr.log](https://github.com/moonlight-stream/moonlight-qt/files/7971540/moonlight-qt-hevs-hdr.log)
Parsed edid TV: [edid_philips.txt](https://github.com/moonlight-stream/moonlight-qt/files/7971549/edid_philips.txt)
Raw edid TV: [edid_philips.dat.zip](https://github.com/moonlight-stream/moonlight-qt/files/7971560/edid_philips.dat.zip)

Log part that seems relevant:
```
00:00:52 - SDL Info (0): Requesting IDR frame on behalf of DR
00:00:52 - SDL Info (0): IDR frame request sent
00:00:52 - SDL Info (0): COLOR_RANGE: YCbCr limited range
00:00:52 - SDL Info (0): COLOR_ENCODING: ITU-R BT.601 YCbCr
00:00:52 - SDL Info (0): Waiting for IDR frame
00:01:03 - SDL Warn (0): HDR_OUTPUT_METADATA is unavailable on this display. Unable to enter HDR mode!
00:01:04 - FFmpeg: [hevc @ 0x24eabb0] Hwaccel V4L2 HEVC stateless V1; devices: /dev/media0,/dev/video19
00:01:04 - SDL Info (0): COLOR_ENCODING: ITU-R BT.2020 YCbCr
00:01:31 - FFmpeg: [hevc @ 0x24eabb0] Hwaccel V4L2 HEVC stateless V1; devices: /dev/media0,/dev/video19
00:01:32 - FFmpeg: [hevc @ 0x24eabb0] Hwaccel V4L2 HEVC stateless V1; devices: /dev/media0,/dev/video19
00:01:32 - SDL Info (0): COLOR_ENCODING: ITU-R BT.601 YCbCr
00:01:32 - SDL Warn (0): HDR_OUTPUT_METADATA is unavailable on this display. Unable to enter HDR mode!
00:01:32 - FFmpeg: [hevc @ 0x24eabb0] Hwaccel V4L2 HEVC stateless V1; devices: /dev/media0,/dev/video19
00:01:32 - SDL Info (0): COLOR_ENCODING: ITU-R BT.2020 YCbCr
00:01:32 - FFmpeg: [hevc @ 0x24eabb0] Hwaccel V4L2 HEVC stateless V1; devices: /dev/media0,/dev/video19
00:01:32 - SDL Info (0): COLOR_ENCODING: ITU-R BT.601 YCbCr
00:01:50 - SDL Warn (0): HDR_OUTPUT_METADATA is unavailable on this display. Unable to enter HDR mode!
00:01:50 - FFmpeg: [hevc @ 0x24eabb0] Hwaccel V4L2 HEVC stateless V1; devices: /dev/media0,/dev/video19
00:01:50 - SDL Warn (0): HDR_OUTPUT_METADATA is unavailable on this display. Unable to enter HDR mode!
00:01:50 - SDL Info (0): COLOR_ENCODING: ITU-R BT.2020 YCbCr
00:01:50 - FFmpeg: [hevc @ 0x24eabb0] Hwaccel V4L2 HEVC stateless V1; devices: /dev/media0,/dev/video19
00:01:50 - SDL Info (0): COLOR_ENCODING: ITU-R BT.601 YCbCr
00:01:50 - FFmpeg: [hevc @ 0x24eabb0] Hwaccel V4L2 HEVC stateless V1; devices: /dev/media0,/dev/video19
00:01:50 - FFmpeg: Failed to alloc 4177920 from dma-heap(fd=39): 12 (Cannot allocate memory)
00:01:50 - FFmpeg: [hevc @ 0x24eabb0] v4l2_request_hevc_start_frame: Failed to get dst buffer
00:01:50 - FFmpeg: [hevc @ 0x24eabb0] Error parsing NAL unit #0.
00:01:50 - SDL Warn (0): avcodec_send_packet() failed: Cannot allocate memory
```

Is there a way to make SDL understand de HDR metadata?

**Comment by veldenb** [comment]
I got it working!

I replaced `dtoverlay=vc4-kms-v3d` with `dtoverlay=vc4-fkms-v3d` (https://github.com/moonlight-stream/moonlight-docs/wiki/Installing-Moonlight-Qt-on-Raspberry-Pi-4).
For HDR that should NOT be done.

The default config.txt only needs `dtoverlay=rpivid-v4l2` to be added and the rest should be left at its default values.
Now the display switches to HDR mode when HDR content is loaded :)

**Comment by veldenb** [comment]
@cgutman HDR is running pretty well on the Pi4 but I do encounter a bug. When using Steam in Big Picture mode the stream switches between SDR and HDR and I run into the following case running the latest code from master:
- Steam Big picture starts in SDR: ok
- Start a game in HDR: ok
- Quit the game and back to Steam Big Picture in SDR: ok
- Start some other game in HDR: the stream freezes and I can only quit moonlight using a kill -9 command

Part of the log:
```
00:10:02 - SDL Info (0): COLOR_ENCODING: ITU-R BT.601 YCbCr
00:10:02 - FFmpeg: [hevc @ 0x2cd0230] Hwaccel V4L2 HEVC stateless V1; devices: /dev/media0,/dev/video19
00:10:02 - SDL Info (0): Set display HDR mode: enabled
00:10:03 - SDL Info (0): Video decode unit queue overflow
00:10:03 - SDL Info (0): IDR frame request sent
00:10:03 - SDL Info (0): Video decode unit queue overflow
00:10:03 - SDL Info (0): IDR frame request sent
00:10:03 - SDL Info (0): Waiting for IDR frame
00:10:03 - SDL Info (0): Video decode unit queue overflow
00:10:03 - SDL Info (0): IDR frame request sent
00:10:04 - SDL Info (0): Video decode unit queue overflow
00:10:04 - SDL Info (0): IDR frame request sent
00:10:04 - SDL Info (0): Video decode unit queue overflow
00:10:04 - SDL Info (0): IDR frame request sent
00:10:04 - SDL Info (0): Waiting for IDR frame
00:10:04 - SDL Info (0): Video decode unit queue overflow
00:10:04 - SDL Info (0): IDR frame request sent
00:10:04 - FFmpeg: [hevc @ 0x2cd0230] mediabufs_poll_cb: Timeout
00:10:04 - FFmpeg: Unable to reinit media request: Device or resource busy
00:10:04 - SDL Info (0): Video decode unit queue overflow
00:10:04 - SDL Info (0): IDR frame request sent
```
For the complete log see 
[moonlight-qt.log](https://github.com/moonlight-stream/moonlight-qt/files/8112705/moonlight-qt.log)
You can ignore the Nintendo Pro game-pad reconnecting, that's a LibreELEC Bluetooth issue.


PS: My TV detects 8-bit colors in HDR-mode, are 10-bit colors also possible or is this a hardware limitation on the Pi?

**Comment by cgutman** [comment]
Released in v4.0.0. 

>Start some other game in HDR: the stream freezes and I can only quit moonlight using a kill -9 command

I don't think this is a Moonlight problem. It seems like a problem with the HEVC codec/driver. I can't even reproduce it anymore with the latest Raspberry Pi OS updates.

Sometimes it will get wedged in a state where it drops a frame every few frames (likely due to memory fragmentation or something similar), but this doesn't result in a full hang.

> PS: My TV detects 8-bit colors in HDR-mode, are 10-bit colors also possible or is this a hardware limitation on the Pi?

No clue, but we are giving the DRM driver 10-bit YUV buffers, so it's out of our hands at that point.

**Comment by veldenb** [comment]
I'll check it again when LibreELEC releases a new kernel version.
Great job adding the HDR :)
